### PR TITLE
Server replication

### PR DESCRIPTION
**Pull request by cassianojaeger**

### CODE REVIEW

File: Instant-Messenger/Makefile [hunk 1]
```diff
@@ -33,7 +33,7 @@ Uuid.o:	$(SRC_DIR)/util/Uuid.cpp
 	${CC} -c $(SRC_DIR)/util/Uuid.cpp
 
 run-client:
-	$(BUILD_DIR)/client.app JohnnyUser1 SampleRoom1 127.0.0.1 $(FE_CLIENTS_PORT)
+	$(BUILD_DIR)/client.app cassianoTest CASSIANOFODATOTAL 127.0.0.1 $(FE_CLIENTS_PORT)
```
Comment (Andy9822) [inline Instant-Messenger/Makefile]: Keep it as before

File: Instant-Messenger/messages/CASSIANOFODATOTAL.csv [hunk 2]
```diff
@@ -0,0 +1,6 @@
+cassianoTestoie1605923883
+cassianoTestREPLICOU SERA?1605923896
+cassianoTestvai replicar ou nao vai caralhooo1605923907
+cassianoTestoie1605923917
+cassianoTestoi1605923918
+cassianoTestola1605923918
```
Comment (Andy9822) [inline Instant-Messenger/messages/CASSIANOFODATOTAL.csv]: No need to update .csv

File: Instant-Messenger/messages/TESTE2.csv [hunk 3]
```diff
@@ -0,0 +1,3 @@
+Joaoreplicacao VOANDO GARAI1605923208
```
Comment (Andy9822) [inline Instant-Messenger/messages/TESTE2.csv]: No need to update .csv

File: Instant-Messenger/src/app/app_server.cpp [hunk 4]
```diff
@@ -30,16 +30,18 @@ void capture_signals()
...
 }
 
-void read_args(int argc, char *argv[], int *port, int *maxNumberOfMessagesInHistory)
+void read_args(int argc, char *argv[], int *port, int *maxNumberOfMessagesInHistory, int *rmNumber, bool *isPrimaryServer)
 {	
 	*port = RANDOM_PORT_NUMBER; //Set any available port as default
 	*maxNumberOfMessagesInHistory = DEFAULT_NUMBER_OF_RECORDED_MESSAGES;
 
-	if(argc == 3)
+	if(argc == 5)
```
Comment (Andy9822) [inline Instant-Messenger/src/app/app_server.cpp]: Check this out, maybe we can remove unnecessary args

File: Instant-Messenger/src/proxy_fe/ProxyFE.cpp [hunk 5]
```diff
@@ -69,7 +69,7 @@ void ProxyFE::prepareSocketConnection(int* socket_fd, sockaddr_in* serv_addr)
     }
 
     // Forcefully attaching socket to the port
-    if (setsockopt(*socket_fd, SOL_SOCKET, SO_REUSEADDR | SO_REUSEPORT, &opt, sizeof(opt))) 
+    if (setsockopt(*socket_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)))
```
Comment (Andy9822) [inline Instant-Messenger/src/proxy_fe/ProxyFE.cpp:72]: Use as it was before

Comment (cassianojaeger) [reply Instant-Messenger/src/proxy_fe/ProxyFE.cpp:72]: It does not work on Macos.

Comment (cassianojaeger) [reply Instant-Messenger/src/proxy_fe/ProxyFE.cpp:72]: Period. We need to either use SO_REUSEADDR or SO_REUSEPORT. Based on my researches, SO_REUSEADDR is advised.

File: Instant-Messenger/src/server/server.cpp [hunk 6]
```diff
@@ -30,6 +31,20 @@ namespace server {
...
+        bzero(&(rm_listening_serv_addr.sin_zero), 8);
+
+        rm_listening_socket_fd = 0;
+    }
+
+    void Server::prepareReplicationManager(int rmNumber) {
+        this->setRmNumber(rmNumber);
+        this->connectToRmServers();
+
+        if(this->getRmNumber() > 0) {
```
Comment (Andy9822) [inline Instant-Messenger/src/server/server.cpp:45]: ?

File: Instant-Messenger/src/server/server.cpp [hunk 7]
```diff
@@ -312,4 +359,199 @@ namespace server {
...
+        rm_listening_serv_addr.sin_port = htons(rmListeningPort);
+
+        // Create socket file descriptor
+        if ((rm_listening_socket_fd = socket(AF_INET, SOCK_STREAM, 0)) <= 0) {
+            cout << "ERROR opening socket\n" << endl;
+            exit(1);
+        }
+
+        // Forcefully attaching socket to the port
+        if (setsockopt(rm_listening_socket_fd, SOL_SOCKET, SO_REUSEADDR, &opt, sizeof(opt)))
```
Comment (Andy9822) [inline Instant-Messenger/src/server/server.cpp:417]: Use it as defined by Alberto and previous etapa

File: Instant-Messenger/src/server/server.cpp (hunk 6 above)
Comment (cassianojaeger) [reply Instant-Messenger/src/server/server.cpp:45]: First RM server should not create listener sockets.

File: Instant-Messenger/src/server/server.cpp (hunk 7 above)
Comment (cassianojaeger) [reply Instant-Messenger/src/server/server.cpp:417]: Previous etapa é foda aeuhueahuea

Comment (cassianojaeger) [reply Instant-Messenger/src/server/server.cpp:417]: Same issue as above discussion. No support for MacOS systems.

File: Instant-Messenger/src/server/server.cpp [hunk 8]
```diff
@@ -312,4 +359,199 @@ namespace server {
...
 
+
+    void Server::connectToRmServers() {
+        for(int connectMachineNumber = rmNumber; connectMachineNumber < MAX_RM; connectMachineNumber++) {
+            int *connectSocket = (int *) calloc(1, sizeof(int *));
+            char ip_address[10] = "127.0.0.1";
+            struct sockaddr_in rm_connect_socket_addr{};
+
+            rm_connect_socket_addr.sin_family = AF_INET;
+            rm_connect_socket_addr.sin_port = htons(RM_BASE_PORT_NUMBER + connectMachineNumber + 1); // port is base port + next machine RM number (rmNumber+1)
```
Comment (Andy9822) [inline Instant-Messenger/src/server/server.cpp:370]: Check this out to maybe flexibilize it

Comment (cassianojaeger) [reply Instant-Messenger/src/server/server.cpp:370]: Already discussed with Gabriel. Lets first make everything work and then we can tackle enhancements and refactorings.

**Review by Andy9822 — COMMENTED** [review]
Solve comments

**Review by Andy9822 — CHANGES_REQUESTED** [review]
Solve comments

**Comment by cassianojaeger** [comment]
> Solve comments

Done.

**Review by Andy9822 — APPROVED** [review]

**Pull request merged**